imagedephi/redact/tiff.py

```diff
@@ -1,83 +1,130 @@
 from __future__ import annotations
 
 from collections.abc import Generator
+import os
 from pathlib import Path
-from typing import TYPE_CHECKING, Type
+from typing import TYPE_CHECKING, Self, Type
+import uuid
 
+from PIL import Image
 import tifftools
 import tifftools.constants
 
-from imagedephi.models.rules import ConcreteMetadataRule, FileFormat, Ruleset
+from imagedephi.models.rules import (
+    ConcreteImageRule,
+    ConcreteMetadataRule,
+    FileFormat,
+    ReplaceImageRule,
+    Ruleset,
+)
 from imagedephi.utils.tiff import get_tiff_tag
 
-from .redaction_plan import FILE_EXTENSION_MAP, RedactionPlan
+from .redaction_plan import RedactionPlan
 
 if TYPE_CHECKING:
-    from tifftools.tifftools import IFD, TiffInfo
+    from tifftools.tifftools import IFD, TagEntry, TiffInfo
 
 
-class TiffBasedMetadataRedactionPlan(RedactionPlan):
+class TiffLikeRedactionPlan(RedactionPlan):
     tiff_info: TiffInfo
 
     def __init__(self, tiff_info: TiffInfo, base_rules: Ruleset, override_rule_set: Ruleset | None):
         self.tiff_info = tiff_info
 
+    @staticmethod
+    def _iter_tiff_tag_entries(
+        ifds: list[IFD],
+        tag_set=tifftools.constants.Tag,
+    ) -> Generator[tuple[tifftools.TiffTag, IFD], None, None]:
+        for ifd in ifds:
+            for tag_id, entry in sorted(ifd["tags"].items()):
+                tag: tifftools.TiffTag = tifftools.constants.get_or_create_tag(
+                    tag_id,
+                    tagSet=tag_set,
+                    datatype=tifftools.Datatype[entry["datatype"]],
+                )
+                if not tag.isIFD():
+                    yield tag, ifd
+                else:
+                    # entry['ifds'] contains a list of lists
+                    # see tifftools.read_tiff
+                    for sub_ifds in entry.get("ifds", []):
+                        yield from TiffLikeRedactionPlan._iter_tiff_tag_entries(
+                            sub_ifds, tag.get("tagset")
+                        )
+
     @classmethod
-    def get_all_subclasses(cls) -> Generator[Type[TiffBasedMetadataRedactionPlan], None, None]:
+    def get_all_subclasses(cls) -> Generator[Type[Self], None, None]:
         for subclass in cls.__subclasses__():
             yield from subclass.get_all_subclasses()
             yield subclass
 
     @classmethod
     def build(
-        cls, image_path: Path, base_rules: Ruleset, override_rules: Ruleset | None = None
-    ) -> TiffBasedMetadataRedactionPlan:
-        file_extension = FILE_EXTENSION_MAP[image_path.suffix]
+        cls,
+        tiff_info: TiffInfo,
+        file_format: FileFormat,
+        base_rules: Ruleset,
+        override_rules: Ruleset | None = None,
+    ) -> Self:
```

Comment:
Previously, each redaction plan opened the file independently. Now the `tiff_info` gets passed directly so multiple plans can operate on the same image data.

Reply:
This is no longer relevant. Instead of having multiple redaction plans, one for metadata and one for images, they are rolled into one.